### DISCUSSION TITLE
Global exception handling

### DISCUSSION THREAD

**Discussion by achobanov** [PrismLibrary/Prism#2568] [WPF]
I'm trying to achieve a global uniform exception handling, as opposed to having to use `try-catch` blocks everywhere. I've seen the `DispatcherUnhandledException` event as a suggestion on a WPF-based solution. I've used it like this:
```c#
public partial class App : PrismApplication
{

    public App()
    {
        this.DispatcherUnhandledException += this.Application_UnhandledException;
    }

    private void Application_UnhandledException(object sender, DispatcherUnhandledExceptionEventArgs e)
    {
        var eventAggregator = ServiceProvider.GetService<IEventAggregator>();
        eventAggregator.GetEvent<ErrorEvent>().Publish(e.Exception.Message);
        e.Handled = true;
    }
...
}
```
However that does not seem to work. My exception is thrown in the Main Thread, as can be seen here:
![image](https://user-images.githubusercontent.com/1363532/134341460-fe08ad03-ca94-482a-b88e-0674050c64a3.png)

Both - ViewModel and exception throw happen on the same thread. 

Is there something else that I need to to in the context of Prism to make this work?

**Comment by sigmarsson** [comment]
Don't do it. Handle the respective exceptions inside their context. You can even introduce your own [Trace Listener](https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/trace-listeners) to track events.

The Global one shall be your very last resort.

**Reply by achobanov** [reply]
Care to elaborate? Centralised handling of exceptions is a very good way to structure validations and fault-handling. Traces may can account for logging, but I don't see how it is better for my application to perform ui actions (as redirecting to an error view) based on traces is better than a simple global exception handling. Even if it's possible (haven't used them myself) it sounds wrong to me. Handling custom exception globally on the other hand is semantically clean and effective

**Comment by brianlagunas** [comment]
Just so we are all on the same page, this has nothing to do with Prism.

These are some events you may want to handle to capture errors within you WPF application:
- Dispatcher.UnhandledException
- Application.DispatcherUnhandledException
- AppDomain.UnhandledException
- TaskScheduler.UnobservedTaskException

**Reply by achobanov** [reply]
Thanks, unfortunately the only handler that works in my case is AppDomain.UnhandledException. That one however does not expose functionality to prevent the exception from exiting the application. It seems I'll have to try-catch every call that could throw.

**Reply by brianlagunas** [reply]
Well, global error handling like this is usually a bad idea anyways. This should only be used for instances that you have an exception that you have not already handled.  Especially since this specific event is not raised for exceptions that corrupt the state of the process, such as stack overflows or access violations

**Reply by achobanov** [reply]
Could you please point me to an article or something that goes in depth as to why it is a bad idea? I'm coming from a web background, mostly on .net 4.7, Core and 5. Having global Exception handling is the go-to for validation and/or uniform responses. All I am trying to do is  have a single place where I'm handling my custom-made exceptions.

**Reply by brianlagunas** [reply]
A common handler is different than trying to have a global event. Seems you have some research to do :)

**Reply by achobanov** [reply]
Yea, I know. Common handling is having a piece of code that handles exceptions in an uniform way. Using global event is a means to run this code, without having to call `HandleException(() => action());` on any line that is expected to throw exceptions. W/E, I would love to be able to spend a whole week on this topic, but I've already have a solution, albeit not a pretty one, and I'll roll with it.

**Reply by damonpkuml** [reply]
Can you show your solution ? It will be very helpfull to me .

**Reply by achobanov** [reply]
@damonpkuml I'm sorry to say I still don't have a clean solution for UI frameworks. I was surprized that the same issue creeps into Blazor as well, although they do have an `ErrorBoundary` component which handles some expcetions, but not all (I don't know why). I'm puzzled to this day as to why UI frameworks don't seem to include this functionality. Unfortunately both @brianlagunas and @sigmarsson above don't contribute to a useful extent.

My solution (for UI frameworks) is to essentially wrap every call at the *Behind level within a `Action` or `Func` and then execute those delegate types in a try-catch block. You can see the current [solution](https://github.com/Not-Endurance/endurance-management-system/blob/release-witness-v5/src/Judge/EMS.Judge/Services/StatelessExecutor.cs)  I'm using for Prism and a [sample usage](https://github.com/Not-Endurance/endurance-management-system/blob/release-witness-v5/src/Judge/EMS.Judge/Services/Implementations/NavigationService.cs#L33). Keep in mind this repo is legacy. I've done a bit better in my Blazor project with [SafeHelper](https://github.com/Not-Endurance/not-timing-system/blob/develop/src/Not/Not/Safe/SafeHelper.cs), but it's essentially the same thing.
